GCC suggest-override warnings

Complete list of warnings when Boost 1.72 is built with GCC 7.5 with `-Wsuggest-override` added to `cxxflags`. Duplicated warnings from same location are omitted:

<pre>
./boost/spirit/home/classic/core/non_terminal/impl/grammar.ipp:173:9: warning: ‘int boost::spirit::classic::impl::grammar_helper<GrammarT, DerivedT, ScannerT>::undefine(boost::spirit::classic::impl::grammar_helper<GrammarT, DerivedT, ScannerT>::grammar_t*) [with GrammarT = boost::spirit::classic::grammar<boost::wave::grammars::chlit_grammar, boost::spirit::classic::closure_context<boost::wave::grammars::closures::chlit_closure> >; DerivedT = boost::wave::grammars::chlit_grammar; ScannerT = boost::spirit::classic::scanner<>; boost::spirit::classic::impl::grammar_helper<GrammarT, DerivedT, ScannerT>::grammar_t = boost::spirit::classic::grammar<boost::wave::grammars::chlit_grammar, boost::spirit::classic::closure_context<boost::wave::grammars::closures::chlit_closure> >]’ can be marked override [-Wsuggest-override]
./boost/spirit/home/classic/core/non_terminal/impl/rule.ipp:239:13: warning: ‘typename boost::spirit::classic::match_result<ScannerT, AttrT>::type boost::spirit::classic::impl::concrete_parser<ParserT, ScannerT, AttrT>::do_parse_virtual(const ScannerT&) const [with ParserT = boost::spirit::classic::action<boost::spirit::classic::rule<boost::spirit::classic::scanner<__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >, boost::spirit::classic::scanner_policies<> >, boost::spirit::classic::nil_t, boost::spirit::classic::nil_t>, boost::archive::xml::append_string<std::__cxx11::basic_string<char>, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > > >; ScannerT = boost::spirit::classic::scanner<__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >, boost::spirit::classic::scanner_policies<> >; AttrT = boost::spirit::classic::nil_t; typename boost::spirit::classic::match_result<ScannerT, AttrT>::type = boost::spirit::classic::match<boost::spirit::classic::nil_t>]’ can be marked override [-Wsuggest-override]
./boost/spirit/home/classic/core/non_terminal/impl/rule.ipp:245:13: warning: ‘boost::spirit::classic::impl::abstract_parser<ScannerT, AttrT>* boost::spirit::classic::impl::concrete_parser<ParserT, ScannerT, AttrT>::clone() const [with ParserT = boost::spirit::classic::action<boost::spirit::classic::rule<boost::spirit::classic::scanner<__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >, boost::spirit::classic::scanner_policies<> >, boost::spirit::classic::nil_t, boost::spirit::classic::nil_t>, boost::archive::xml::append_string<std::__cxx11::basic_string<char>, __gnu_cxx::__normal_iterator<const char*, std::__cxx11::basic_string<char> > > >; ScannerT = boost::spirit::classic::scanner<__gnu_cxx::__normal_iterator<char*, std::__cxx11::basic_string<char> >, boost::spirit::classic::scanner_policies<> >; AttrT = boost::spirit::classic::nil_t]’ can be marked override [-Wsuggest-override]
./boost/spirit/home/classic/iterator/multi_pass.hpp:156:5: warning: ‘virtual const char* boost::spirit::classic::multi_pass_policies::illegal_backtracking::what() const’ can be marked override [-Wsuggest-override]
./boost/spirit/home/support/detail/hold_any.hpp:52:29: warning: ‘virtual const char* boost::spirit::bad_any_cast::what() const’ can be marked override [-Wsuggest-override]
./boost/spirit/home/support/iterators/detail/buf_id_check_policy.hpp:29:21: warning: ‘virtual const char* boost::spirit::iterator_policies::illegal_backtracking::what() const’ can be marked override [-Wsuggest-override]
</pre>

`BOOST_OVERRIDE` was introduced in https://github.com/boostorg/config/commit/ffe4e0f5a448578cce14e3eed0cf7163333a81d9.

--- a/include/boost/spirit/home/classic/core/non_terminal/impl/grammar.ipp
+++ b/include/boost/spirit/home/classic/core/non_terminal/impl/grammar.ipp
@@ -170,7 +170,7 @@ struct grammar_definition
         }
 
         int
-        undefine(grammar_t* target_grammar)
+        undefine(grammar_t* target_grammar) BOOST_OVERRIDE
         {
             typename grammar_t::object_id id = target_grammar->get_object_id();
 

--- a/include/boost/spirit/home/classic/core/non_terminal/impl/object_with_id.ipp
+++ b/include/boost/spirit/home/classic/core/non_terminal/impl/object_with_id.ipp
@@ -102,7 +102,7 @@ BOOST_SPIRIT_CLASSIC_NAMESPACE_BEGIN
 #ifdef BOOST_SPIRIT_THREADSAFE
             boost::unique_lock<boost::mutex> lock(mutex);
 #endif
-            if (free_ids.size())
+            if (!free_ids.empty())
             {
                 object_id id = *free_ids.rbegin();
                 free_ids.pop_back();

--- a/include/boost/spirit/home/classic/core/non_terminal/impl/rule.ipp
+++ b/include/boost/spirit/home/classic/core/non_terminal/impl/rule.ipp
@@ -232,16 +232,16 @@ BOOST_SPIRIT_CLASSIC_NAMESPACE_BEGIN
         struct concrete_parser : abstract_parser<ScannerT, AttrT>
         {
             concrete_parser(ParserT const& p_) : p(p_) {}
-            virtual ~concrete_parser() {}
+            ~concrete_parser() BOOST_OVERRIDE {}
 
-            virtual typename match_result<ScannerT, AttrT>::type
-            do_parse_virtual(ScannerT const& scan) const
+            typename match_result<ScannerT, AttrT>::type
+            do_parse_virtual(ScannerT const& scan) const BOOST_OVERRIDE
             {
                 return p.parse(scan);
             }
 
-            virtual abstract_parser<ScannerT, AttrT>*
-            clone() const
+            abstract_parser<ScannerT, AttrT>*
+            clone() const BOOST_OVERRIDE
             {
                 return new concrete_parser(p);
             }

--- a/include/boost/spirit/home/classic/iterator/fixed_size_queue.hpp
+++ b/include/boost/spirit/home/classic/iterator/fixed_size_queue.hpp
@@ -9,9 +9,9 @@
 #ifndef BOOST_SPIRIT_CLASSIC_ITERATOR_FIXED_SIZE_QUEUE_HPP
 #define BOOST_SPIRIT_CLASSIC_ITERATOR_FIXED_SIZE_QUEUE_HPP
 
+#include <cstddef>
 #include <cstdlib>
 #include <iterator>
-#include <cstddef>
 
 #include <boost/spirit/home/classic/namespace.hpp>
 #include <boost/spirit/home/classic/core/assert.hpp> // for BOOST_SPIRIT_ASSERT

--- a/include/boost/spirit/home/classic/iterator/multi_pass.hpp
+++ b/include/boost/spirit/home/classic/iterator/multi_pass.hpp
@@ -150,10 +150,10 @@ class BOOST_SYMBOL_VISIBLE illegal_backtracking : public std::exception
 public:
 
     illegal_backtracking() BOOST_NOEXCEPT_OR_NOTHROW {}
-    ~illegal_backtracking() BOOST_NOEXCEPT_OR_NOTHROW {}
+    ~illegal_backtracking() BOOST_NOEXCEPT_OR_NOTHROW BOOST_OVERRIDE {}
 
-    virtual const char*
-    what() const BOOST_NOEXCEPT_OR_NOTHROW
+    const char*
+    what() const BOOST_NOEXCEPT_OR_NOTHROW BOOST_OVERRIDE
     { return "BOOST_SPIRIT_CLASSIC_NS::illegal_backtracking"; }
 };
 
@@ -163,7 +163,7 @@ public:
 // This policy is most effective when used together with the std_deque
 // StoragePolicy.
 // If used with the fixed_size_queue StoragePolicy, it will not detect
-// iterator derefereces that are out of the range of the queue.
+// iterator dereferences that are out of the range of the queue.
 ///////////////////////////////////////////////////////////////////////////////
 class buf_id_check
 {
@@ -1296,5 +1296,3 @@ BOOST_SPIRIT_CLASSIC_NAMESPACE_END
 }} // namespace BOOST_SPIRIT_CLASSIC_NS
 
 #endif // BOOST_SPIRIT_ITERATOR_MULTI_PASS_HPP
-
-

--- a/include/boost/spirit/home/classic/iterator/position_iterator.hpp
+++ b/include/boost/spirit/home/classic/iterator/position_iterator.hpp
@@ -293,7 +293,7 @@ private:
 
 protected:
 
-    void newline(void)
+    void newline()
     {}
 
     ForwardIteratorT _end;
@@ -396,13 +396,13 @@ public:
         return *this;
     }
 
-    ForwardIteratorT get_currentline_begin(void) const
+    ForwardIteratorT get_currentline_begin() const
     { return _startline; }
 
-    ForwardIteratorT get_currentline_end(void) const
+    ForwardIteratorT get_currentline_end() const
     { return get_endline(); }
 
-    std::basic_string<value_type> get_currentline(void) const
+    std::basic_string<value_type> get_currentline() const
     {
         return std::basic_string<value_type>
             (get_currentline_begin(), get_currentline_end());
@@ -424,7 +424,7 @@ protected:
         return endline;
     }
 
-    void newline(void)
+    void newline()
     { _startline = this->base(); }
 };
 

--- a/include/boost/spirit/home/qi/detail/expectation_failure.hpp
+++ b/include/boost/spirit/home/qi/detail/expectation_failure.hpp
@@ -24,7 +24,7 @@ namespace boost { namespace spirit { namespace qi {
             : std::runtime_error("boost::spirit::qi::expectation_failure")
             , first(first_), last(last_), what_(what)
         {}
-        ~expectation_failure() BOOST_NOEXCEPT_OR_NOTHROW {}
+        ~expectation_failure() BOOST_NOEXCEPT_OR_NOTHROW BOOST_OVERRIDE {}
 
         Iterator first;
         Iterator last;

--- a/include/boost/spirit/home/support/detail/hold_any.hpp
+++ b/include/boost/spirit/home/support/detail/hold_any.hpp
@@ -27,10 +27,10 @@
 #include <boost/assert.hpp>
 #include <boost/core/typeinfo.hpp>
 
-#include <stdexcept>
-#include <typeinfo>
 #include <algorithm>
 #include <iosfwd>
+#include <stdexcept>
+#include <typeinfo>
 
 ///////////////////////////////////////////////////////////////////////////////
 #if BOOST_WORKAROUND(BOOST_MSVC, >= 1400)
@@ -49,7 +49,10 @@ namespace boost { namespace spirit
           : from(src.name()), to(dest.name())
         {}
 
-        virtual const char* what() const BOOST_NOEXCEPT_OR_NOTHROW { return "bad any cast"; }
+        const char* what() const BOOST_NOEXCEPT_OR_NOTHROW BOOST_OVERRIDE
+        { 
+            return "bad any cast";
+        }
 
         const char* from;
         const char* to;

--- a/include/boost/spirit/home/support/info.hpp
+++ b/include/boost/spirit/home/support/info.hpp
@@ -132,7 +132,7 @@ namespace boost { namespace spirit
 
         void element(string const& tag, string const& value, int /*depth*/) const
         {
-            if (value == "")
+            if (value.empty())
                 out << '<' << tag << '>';
             else
                 out << '"' << value << '"';

--- a/include/boost/spirit/home/support/iterators/detail/buf_id_check_policy.hpp
+++ b/include/boost/spirit/home/support/iterators/detail/buf_id_check_policy.hpp
@@ -24,9 +24,9 @@ namespace boost { namespace spirit { namespace iterator_policies
     {
     public:
         illegal_backtracking() BOOST_NOEXCEPT_OR_NOTHROW {}
-        ~illegal_backtracking() BOOST_NOEXCEPT_OR_NOTHROW {}
+        ~illegal_backtracking() BOOST_NOEXCEPT_OR_NOTHROW BOOST_OVERRIDE {}
 
-        char const* what() const BOOST_NOEXCEPT_OR_NOTHROW
+        char const* what() const BOOST_NOEXCEPT_OR_NOTHROW BOOST_OVERRIDE
         { 
             return "boost::spirit::multi_pass::illegal_backtracking"; 
         }

